cogip/sensor.py

```diff
@@ -8,46 +9,61 @@
 
 class Sensor(QtCore.QObject):
 
+    # Class attribute recording all entities that should be detected
+    obstacles = []
+
+    # Class attribute recording all sensors
+    # Each added obstacle must be registered in all sensors
+    all_sensors = []
+
     def __init__(
             self,
             asset_entity: AssetEntity,
             name: str,
             origin_x: int,
             origin_y: int,
+            origin_z: int,
+            direction_x: int,
+            direction_y: int,
+            direction_z: int,
             impact_radius: float = 50,
-            impact_color: QtCore.Qt.red = QtCore.Qt.red):
+            impact_color: QtCore.Qt.GlobalColor = QtCore.Qt.red):
         super(Sensor, self).__init__()
 
+        Sensor.all_sensors.append(self)
+
         self.asset_entity = asset_entity
         self.name = name
         self.distance = None
 
-        direction_x = 1.0 if origin_x > 0 else -1.0 if origin_x < 0 else 0.0
-        direction_y = 1.0 if origin_y > 0 else -1.0 if origin_y < 0 else 0.0
-
         self.ray_caster = Qt3DRender.QRayCaster()
-        self.ray_caster.setEnabled(True)
-        self.ray_caster.setLength(0)  # Infinite
+        self.ray_caster.setEnabled(False)
+        self.ray_caster.setLength(2000)  # Infinite
         self.ray_caster.setRunMode(Qt3DRender.QAbstractRayCaster.Continuous)
-        self.ray_caster.setOrigin(QtGui.QVector3D(float(origin_x), float(origin_y), 60))
-        self.ray_caster.setDirection(QtGui.QVector3D(direction_x, direction_y, 0))
-
+        # self.ray_caster.setRunMode(Qt3DRender.QAbstractRayCaster.SingleShot)
+        self.ray_caster.setFilterMode(Qt3DRender.QAbstractRayCaster.AcceptAnyMatchingLayers)
+        self.ray_caster.setOrigin(QtGui.QVector3D(float(origin_x), float(origin_y), float(origin_z)))
+        self.ray_caster.setDirection(QtGui.QVector3D(direction_x, direction_y, direction_z))
         self.asset_entity.asset_entity.addComponent(self.ray_caster)
 
+        # Add layers for obstacles already present
+        for obstacle in self.obstacles:
+            self.add_obstacle_layer(obstacle)
+
         # Add impact entity
         self.impact_entity = ImpactEntity(radius=impact_radius, color=impact_color)
         self.impact_entity.setParent(self.asset_entity)
 
         self.timer = QtCore.QTimer()
         self.timer.timeout.connect(self.handle_hits)
-        self.timer.start(50)
+        self.timer.start(100)
```

Comment:
what is this timer for and why changing 50 to 100 ? Is this milliseconds ?

Reply:
It's milliseconds. Reverted also.
Obstacles are detected each time the timer timeout (here every 50 ms).